File: 15. 3Sum/3Sum.go

func threeSum(nums []int) [][]int {
	// Sort the input array for easier management
	sort.Ints(nums)
	n := len(nums)
	result := [][]int{}

	for i := 0; i < n-2; i++ {
		// Skip duplicates
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		low, high := i+1, n-1
		target := -nums[i]

		for low < high {
			sum := nums[low] + nums[high]

			if sum == target {
				result = append(result, []int{nums[i], nums[low], nums[high]})
				// Skip duplicates
				for low < high && nums[low] == nums[low+1] {
					low++
				}
				for low < high && nums[high] == nums[high-1] {
					high--
				}
				low++
				high--
			} else if sum < target {
				low++
			} else {
				high--
			}
		}
	}

	return result
}

File: 15. 3Sum/3Sum.py
def threeSum(nums):
    result = []
    if not nums or len(nums) < 3:
        return result
    
    nums.sort()  # O(n log n)
    for i in range(len(nums) - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue  # Skip duplicates
        left, right = i + 1, len(nums) - 1
        while left < right:
            total = nums[i] + nums[left] + nums[right]
            if total == 0:
                result.append([nums[i], nums[left], nums[right]])
                left += 1
                right -= 1
                while left < right and nums[left] == nums[left - 1]:
                    left += 1  # Skip duplicates
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1  # Skip duplicates
            elif total < 0:
                left += 1
            else:
                right -= 1
    return result